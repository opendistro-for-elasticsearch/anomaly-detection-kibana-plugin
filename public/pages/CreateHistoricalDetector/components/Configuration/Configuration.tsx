/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { EuiSpacer } from '@elastic/eui';
import { FormikProps } from 'formik';
import React from 'react';
import ContentPanel from '../../../../components/ContentPanel/ContentPanel';
import { Detector } from '../../../../models/interfaces';
import {
  HistoricalDetectorFormikValues,
  SAVE_HISTORICAL_DETECTOR_OPTIONS,
} from '../../utils/constants';
import { Timestamp } from './components/Timestamp/Timestamp';
import { Features } from './components/Features/Features';
import { OperationSettings } from './components/OperationSettings/OperationSettings';
import { ExistingDetectors } from './components/ExistingDetectors/ExistingDetectors';

interface ConfigurationProps {
  isEdit: boolean;
  detector: Detector;
  isLoading: boolean;
  formikProps: FormikProps<HistoricalDetectorFormikValues>;
  selectedSaveOption: SAVE_HISTORICAL_DETECTOR_OPTIONS;
  onSaveOptionChange(id: string): void;
}

export function Configuration(props: ConfigurationProps) {
  return (
    <ContentPanel title="Detector settings" titleSize="s">
      {!props.isEdit ? (
        <ExistingDetectors formikProps={props.formikProps} />
      ) : null}
      <Timestamp formikProps={props.formikProps} />
      <EuiSpacer size="l" />
      <Features
        detector={props.detector}
        formikProps={props.formikProps}
        isLoading={props.isLoading}
      />
      <EuiSpacer size="l" />
      <OperationSettings
        formikProps={props.formikProps}
        selectedOption={props.selectedSaveOption}
        onOptionChange={props.onSaveOptionChange}
      />
    </ContentPanel>
  );
}
