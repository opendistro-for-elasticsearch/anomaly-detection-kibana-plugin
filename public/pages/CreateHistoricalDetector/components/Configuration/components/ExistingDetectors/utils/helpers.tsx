/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { get } from 'lodash';
import {
  HistoricalDetectorFormikValues,
  INITIAL_HISTORICAL_DETECTOR_VALUES,
} from '../../../../../utils/constants';
import { FormikProps } from 'formik';
import { Detector } from '../../../../../../../models/interfaces';
import { generateInitialFeatures } from '../../../../../../EditFeatures/utils/helpers';

export function populateDetectorFieldsFromDetector(
  formikProps: FormikProps<HistoricalDetectorFormikValues>,
  detector: Detector
) {
  formikProps.setFieldValue('timeField', detector.timeField);
  formikProps.setFieldValue('featureList', generateInitialFeatures(detector));
  formikProps.setFieldValue(
    'detectionInterval',
    get(detector, 'detectionInterval.period.interval')
  );
}

export function populateDetectorFieldsToInitialValues(
  formikProps: FormikProps<HistoricalDetectorFormikValues>
) {
  formikProps.setFieldValue(
    'timeField',
    INITIAL_HISTORICAL_DETECTOR_VALUES.timeField
  );
  formikProps.setFieldValue(
    'featureList',
    INITIAL_HISTORICAL_DETECTOR_VALUES.featureList
  );
  formikProps.setFieldValue(
    'detectionInterval',
    INITIAL_HISTORICAL_DETECTOR_VALUES.detectionInterval
  );
}

export function untouchDetectorFields(
  formikProps: FormikProps<HistoricalDetectorFormikValues>
) {
  formikProps.setFieldTouched('timeField', false);
  formikProps.setFieldTouched('featureList', false);
  formikProps.setFieldTouched('detectionInterval', false);
}
