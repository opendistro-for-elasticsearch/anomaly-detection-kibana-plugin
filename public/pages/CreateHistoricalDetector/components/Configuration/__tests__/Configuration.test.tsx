/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';
import { Provider } from 'react-redux';
import {
  HashRouter as Router,
  RouteComponentProps,
  Route,
  Switch,
} from 'react-router-dom';
import { render } from '@testing-library/react';
import { Configuration } from '../Configuration';
import configureStore from '../../../../../redux/configureStore';
import { httpClientMock, coreServicesMock } from '../../../../../../test/mocks';
import { CoreServicesContext } from '../../../../../components/CoreServices/CoreServices';
import { Formik } from 'formik';
import { SAVE_HISTORICAL_DETECTOR_OPTIONS } from '../../../utils/constants';

const TITLE_TEXT = 'Detector settings';
const EXISTING_DETECTOR_TEXT = 'Configure using an existing detector';

const renderWithRouter = (isEdit: boolean = false) => ({
  ...render(
    <Provider store={configureStore(httpClientMock)}>
      <Router>
        <Switch>
          <Route
            render={(props: RouteComponentProps) => (
              <CoreServicesContext.Provider value={coreServicesMock}>
                <Formik
                  initialValues={{ featureList: [] }}
                  onSubmit={jest.fn()}
                >
                  {() => (
                    <div>
                      <Configuration
                        formikProps={{
                          values: { featureList: [] },
                          setFieldValue: jest.fn(),
                          setFieldTouched: jest.fn(),
                        }}
                        isEdit={isEdit}
                        detector={{}}
                        isLoading={false}
                        selectedSaveOption={
                          SAVE_HISTORICAL_DETECTOR_OPTIONS.START
                        }
                        onSaveOptionChange={jest.fn()}
                        {...props}
                      />
                    </div>
                  )}
                </Formik>
              </CoreServicesContext.Provider>
            )}
          />
        </Switch>
      </Router>
    </Provider>
  ),
});

describe('<Configuration /> spec', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  test('renders the component in create mode with existing detectors option', () => {
    console.error = jest.fn();
    const { container, getByText, queryByText } = renderWithRouter(true);
    expect(container.firstChild).toMatchSnapshot();
    getByText(TITLE_TEXT);
    expect(queryByText(EXISTING_DETECTOR_TEXT)).toBeNull();
  });

  test('renders the component in edit mode without existing detectors option', () => {
    console.error = jest.fn();
    const { container, getByText, findByText } = renderWithRouter(true);
    expect(container.firstChild).toMatchSnapshot();
    getByText(TITLE_TEXT);
    expect(findByText(EXISTING_DETECTOR_TEXT)).not.toBeNull();
  });
});
