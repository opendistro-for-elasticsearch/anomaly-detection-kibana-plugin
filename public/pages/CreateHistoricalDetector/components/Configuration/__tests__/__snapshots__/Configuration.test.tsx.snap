// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Configuration /> spec renders the component in create mode with existing detectors option 1`] = `
<div>
  <div
    class="euiPanel euiPanel--paddingMedium"
    style="padding: 20px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px;"
    >
      <div
        class="euiFlexItem"
      >
        <h3
          class="euiTitle euiTitle--small"
        >
          Detector settings
        </h3>
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          />
        </div>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          />
        </div>
      </div>
    </div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
    />
    <div
      style="padding: 10px 0px;"
    >
      <div
        class="euiPanel euiPanel--paddingMedium"
        style="padding: 20px;"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
          style="padding: 0px;"
        >
          <div
            class="euiFlexItem"
          >
            <h3
              class="euiTitle euiTitle--small"
            >
              Timestamp
            </h3>
            <div
              class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
            >
              <div
                class="euiFlexItem"
              />
            </div>
          </div>
          <div
            class="euiFlexItem euiFlexItem--flexGrowZero"
          >
            <div
              class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
            >
              <div
                class="euiFlexItem"
              />
            </div>
          </div>
        </div>
        <hr
          class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
        />
        <div
          style="padding: 10px 0px;"
        >
          <div
            class="euiFlexItem"
            style="max-width: 70%;"
          >
            <div
              class="euiFormRow"
              hint="Choose the time field you want to use for time filter."
              id="random_html_id-row"
              title="Timestamp field"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  aria-invalid="false"
                  class="euiFormLabel euiFormRow__label"
                  for="random_html_id"
                >
                  <div
                    style="line-height: 8px;"
                  >
                    <p>
                      Timestamp field
                    </p>
                    <br />
                    <p
                      class="sublabel"
                    >
                      Choose the time field you want to use for time filter.
                    </p>
                  </div>
                </label>
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  class="euiComboBox"
                  role="combobox"
                >
                  <div
                    class="euiFormControlLayout"
                  >
                    <div
                      class="euiFormControlLayout__childrenWrapper"
                    >
                      <div
                        class="euiComboBox__inputWrap euiComboBox__inputWrap--noWrap euiComboBox__inputWrap-isClearable"
                        data-test-subj="comboBoxInput"
                        tabindex="-1"
                      >
                        <p
                          class="euiComboBoxPlaceholder"
                        >
                          Find timestamp
                        </p>
                        <div
                          class="euiComboBox__input"
                          style="font-size: 14px; display: inline-block;"
                        >
                          <input
                            aria-controls=""
                            data-test-subj="comboBoxSearchInput"
                            id="random_html_id"
                            role="textbox"
                            style="box-sizing: content-box; width: 2px;"
                            value=""
                          />
                          <div
                            style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                          />
                        </div>
                      </div>
                      <div
                        class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                      >
                        <button
                          aria-label="Open list of options"
                          class="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
                          data-test-subj="comboBoxToggleListButton"
                          type="button"
                        >
                          <svg
                            aria-hidden="true"
                            class="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                            focusable="false"
                            height="16"
                            role="img"
                            viewBox="0 0 16 16"
                            width="16"
                            xmlns="http://www.w3.org/2000/svg"
                          />
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiSpacer euiSpacer--l"
      />
      <div
        class="euiPanel euiPanel--paddingMedium"
        style="padding: 20px;"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
          style="padding: 0px;"
        >
          <div
            class="euiFlexItem"
          >
            <h3
              class="euiTitle euiTitle--small"
            >
              Features
            </h3>
            <div
              class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
            >
              <div
                class="euiFlexItem"
              >
                <div
                  class="euiText euiText--medium content-panel-subTitle"
                >
                  Specify an index field that you want to find anomalies for by defining features. You can add up to 5 features.
                   
                  <a
                    class="euiLink euiLink--primary"
                    href="https://opendistro.github.io/for-elasticsearch-docs/docs/ad/"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    Learn more 
                    <svg
                      aria-hidden="true"
                      class="euiIcon euiIcon--small euiIcon-isLoading"
                      focusable="false"
                      height="16"
                      role="img"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    />
                  </a>
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiFlexItem euiFlexItem--flexGrowZero"
          >
            <div
              class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
            >
              <div
                class="euiFlexItem"
              />
            </div>
          </div>
        </div>
        <hr
          class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
        />
        <div
          style="padding: 10px 0px;"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionColumn euiFlexGroup--responsive"
            style="margin: 0px;"
          >
            <div
              class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
              style="padding: 12px 24px;"
            >
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <button
                  class="euiButton euiButton--primary"
                  data-test-subj="addFeature"
                  type="button"
                >
                  <span
                    class="euiButtonContent euiButton__content"
                  >
                    <span
                      class="euiButton__text"
                    >
                      Add feature
                    </span>
                  </span>
                </button>
                <div
                  class="euiText euiText--medium content-panel-subTitle"
                >
                  <p>
                    You can add up to
                     
                    5
                     
                    more features.
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiSpacer euiSpacer--l"
      />
      <div
        class="euiPanel euiPanel--paddingMedium"
        style="padding: 20px;"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
          style="padding: 0px;"
        >
          <div
            class="euiFlexItem"
          >
            <h3
              class="euiTitle euiTitle--small"
            >
              Operation settings
            </h3>
            <div
              class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
            >
              <div
                class="euiFlexItem"
              />
            </div>
          </div>
          <div
            class="euiFlexItem euiFlexItem--flexGrowZero"
          >
            <div
              class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
            >
              <div
                class="euiFlexItem"
              />
            </div>
          </div>
        </div>
        <hr
          class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
        />
        <div
          style="padding: 10px 0px;"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionColumn euiFlexGroup--responsive"
            style="margin: 0px;"
          >
            <div
              class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionColumn euiFlexGroup--responsive"
            >
              <div
                class="euiFlexItem"
                style="max-width: 70%;"
              >
                <div
                  class="euiFormRow euiFormRow--fullWidth"
                  hint="Define how often the detector collects data to generate anomalies. Shorter intervals generate more detector results and  therefore require more computing resources."
                  id="random_html_id-row"
                  title="Detection interval"
                >
                  <div
                    class="euiFormRow__labelWrapper"
                  >
                    <label
                      aria-invalid="false"
                      class="euiFormLabel euiFormRow__label"
                      for="random_html_id"
                    >
                      <div
                        style="line-height: 8px;"
                      >
                        <p>
                          Detection interval
                        </p>
                        <br />
                        <p
                          class="sublabel"
                        >
                          Define how often the detector collects data to generate anomalies. Shorter intervals generate more detector results and  therefore require more computing resources.
                        </p>
                      </div>
                    </label>
                  </div>
                  <div
                    class="euiFormRow__fieldWrapper"
                  >
                    <div
                      class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
                      id="random_html_id"
                    >
                      <div
                        class="euiFlexItem euiFlexItem--flexGrowZero"
                      >
                        <div
                          class="euiFormControlLayout"
                        >
                          <div
                            class="euiFormControlLayout__childrenWrapper"
                          >
                            <input
                              class="euiFieldNumber"
                              data-test-subj="detectionInterval"
                              id="detectionInterval"
                              name="detectionInterval"
                              placeholder="Detector interval"
                              type="number"
                              value=""
                            />
                          </div>
                        </div>
                      </div>
                      <div
                        class="euiFlexItem"
                      >
                        <div
                          class="euiText euiText--medium"
                        >
                          <p
                            class="minutes"
                          >
                            minutes
                          </p>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="euiFlexItem"
              >
                <div
                  class="euiCheckbox"
                >
                  <input
                    checked=""
                    class="euiCheckbox__input"
                    id="runDetectorAutomaticallyCheckbox"
                    type="checkbox"
                  />
                  <div
                    class="euiCheckbox__square"
                  />
                  <label
                    class="euiCheckbox__label"
                    for="runDetectorAutomaticallyCheckbox"
                  >
                    Run detector automatically after creating
                  </label>
                </div>
              </div>
              <div
                class="euiFlexItem"
                style="margin-top: -8px;"
              >
                <p
                  class="sublabel"
                >
                   
                  We recommend running this detector during low traffic times.
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Configuration /> spec renders the component in edit mode without existing detectors option 1`] = `
<div>
  <div
    class="euiPanel euiPanel--paddingMedium"
    style="padding: 20px;"
  >
    <div
      class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      style="padding: 0px;"
    >
      <div
        class="euiFlexItem"
      >
        <h3
          class="euiTitle euiTitle--small"
        >
          Detector settings
        </h3>
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          />
        </div>
      </div>
      <div
        class="euiFlexItem euiFlexItem--flexGrowZero"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem"
          />
        </div>
      </div>
    </div>
    <hr
      class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
    />
    <div
      style="padding: 10px 0px;"
    >
      <div
        class="euiPanel euiPanel--paddingMedium"
        style="padding: 20px;"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
          style="padding: 0px;"
        >
          <div
            class="euiFlexItem"
          >
            <h3
              class="euiTitle euiTitle--small"
            >
              Timestamp
            </h3>
            <div
              class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
            >
              <div
                class="euiFlexItem"
              />
            </div>
          </div>
          <div
            class="euiFlexItem euiFlexItem--flexGrowZero"
          >
            <div
              class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
            >
              <div
                class="euiFlexItem"
              />
            </div>
          </div>
        </div>
        <hr
          class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
        />
        <div
          style="padding: 10px 0px;"
        >
          <div
            class="euiFlexItem"
            style="max-width: 70%;"
          >
            <div
              class="euiFormRow"
              hint="Choose the time field you want to use for time filter."
              id="random_html_id-row"
              title="Timestamp field"
            >
              <div
                class="euiFormRow__labelWrapper"
              >
                <label
                  aria-invalid="false"
                  class="euiFormLabel euiFormRow__label"
                  for="random_html_id"
                >
                  <div
                    style="line-height: 8px;"
                  >
                    <p>
                      Timestamp field
                    </p>
                    <br />
                    <p
                      class="sublabel"
                    >
                      Choose the time field you want to use for time filter.
                    </p>
                  </div>
                </label>
              </div>
              <div
                class="euiFormRow__fieldWrapper"
              >
                <div
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  class="euiComboBox"
                  role="combobox"
                >
                  <div
                    class="euiFormControlLayout"
                  >
                    <div
                      class="euiFormControlLayout__childrenWrapper"
                    >
                      <div
                        class="euiComboBox__inputWrap euiComboBox__inputWrap--noWrap euiComboBox__inputWrap-isClearable"
                        data-test-subj="comboBoxInput"
                        tabindex="-1"
                      >
                        <p
                          class="euiComboBoxPlaceholder"
                        >
                          Find timestamp
                        </p>
                        <div
                          class="euiComboBox__input"
                          style="font-size: 14px; display: inline-block;"
                        >
                          <input
                            aria-controls=""
                            data-test-subj="comboBoxSearchInput"
                            id="random_html_id"
                            role="textbox"
                            style="box-sizing: content-box; width: 2px;"
                            value=""
                          />
                          <div
                            style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                          />
                        </div>
                      </div>
                      <div
                        class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                      >
                        <button
                          aria-label="Open list of options"
                          class="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
                          data-test-subj="comboBoxToggleListButton"
                          type="button"
                        >
                          <svg
                            aria-hidden="true"
                            class="euiIcon euiIcon--medium euiFormControlLayoutCustomIcon__icon"
                            focusable="false"
                            height="16"
                            role="img"
                            viewBox="0 0 16 16"
                            width="16"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M13.069 5.157L8.384 9.768a.546.546 0 01-.768 0L2.93 5.158a.552.552 0 00-.771 0 .53.53 0 000 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 000-.76.552.552 0 00-.771 0z"
                              fill-rule="non-zero"
                            />
                          </svg>
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiSpacer euiSpacer--l"
      />
      <div
        class="euiPanel euiPanel--paddingMedium"
        style="padding: 20px;"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
          style="padding: 0px;"
        >
          <div
            class="euiFlexItem"
          >
            <h3
              class="euiTitle euiTitle--small"
            >
              Features
            </h3>
            <div
              class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
            >
              <div
                class="euiFlexItem"
              >
                <div
                  class="euiText euiText--medium content-panel-subTitle"
                >
                  Specify an index field that you want to find anomalies for by defining features. You can add up to 5 features.
                   
                  <a
                    class="euiLink euiLink--primary"
                    href="https://opendistro.github.io/for-elasticsearch-docs/docs/ad/"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    Learn more 
                    <svg
                      aria-hidden="true"
                      class="euiIcon euiIcon--small"
                      focusable="false"
                      height="16"
                      role="img"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M13 8.5a.5.5 0 111 0V12a2 2 0 01-2 2H4a2 2 0 01-2-2V4a2 2 0 012-2h3.5a.5.5 0 010 1H4a1 1 0 00-1 1v8a1 1 0 001 1h8a1 1 0 001-1V8.5zm-5.12.339a.5.5 0 11-.706-.707L13.305 2H10.5a.5.5 0 110-1H14a1 1 0 011 1v3.5a.5.5 0 11-1 0V2.72L7.88 8.838z"
                      />
                    </svg>
                  </a>
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiFlexItem euiFlexItem--flexGrowZero"
          >
            <div
              class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
            >
              <div
                class="euiFlexItem"
              />
            </div>
          </div>
        </div>
        <hr
          class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
        />
        <div
          style="padding: 10px 0px;"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionColumn euiFlexGroup--responsive"
            style="margin: 0px;"
          >
            <div
              class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
              style="padding: 12px 24px;"
            >
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <button
                  class="euiButton euiButton--primary"
                  data-test-subj="addFeature"
                  type="button"
                >
                  <span
                    class="euiButtonContent euiButton__content"
                  >
                    <span
                      class="euiButton__text"
                    >
                      Add feature
                    </span>
                  </span>
                </button>
                <div
                  class="euiText euiText--medium content-panel-subTitle"
                >
                  <p>
                    You can add up to
                     
                    5
                     
                    more features.
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiSpacer euiSpacer--l"
      />
      <div
        class="euiPanel euiPanel--paddingMedium"
        style="padding: 20px;"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
          style="padding: 0px;"
        >
          <div
            class="euiFlexItem"
          >
            <h3
              class="euiTitle euiTitle--small"
            >
              Operation settings
            </h3>
            <div
              class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
            >
              <div
                class="euiFlexItem"
              />
            </div>
          </div>
          <div
            class="euiFlexItem euiFlexItem--flexGrowZero"
          >
            <div
              class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
            >
              <div
                class="euiFlexItem"
              />
            </div>
          </div>
        </div>
        <hr
          class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
        />
        <div
          style="padding: 10px 0px;"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionColumn euiFlexGroup--responsive"
            style="margin: 0px;"
          >
            <div
              class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionColumn euiFlexGroup--responsive"
            >
              <div
                class="euiFlexItem"
                style="max-width: 70%;"
              >
                <div
                  class="euiFormRow euiFormRow--fullWidth"
                  hint="Define how often the detector collects data to generate anomalies. Shorter intervals generate more detector results and  therefore require more computing resources."
                  id="random_html_id-row"
                  title="Detection interval"
                >
                  <div
                    class="euiFormRow__labelWrapper"
                  >
                    <label
                      aria-invalid="false"
                      class="euiFormLabel euiFormRow__label"
                      for="random_html_id"
                    >
                      <div
                        style="line-height: 8px;"
                      >
                        <p>
                          Detection interval
                        </p>
                        <br />
                        <p
                          class="sublabel"
                        >
                          Define how often the detector collects data to generate anomalies. Shorter intervals generate more detector results and  therefore require more computing resources.
                        </p>
                      </div>
                    </label>
                  </div>
                  <div
                    class="euiFormRow__fieldWrapper"
                  >
                    <div
                      class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
                      id="random_html_id"
                    >
                      <div
                        class="euiFlexItem euiFlexItem--flexGrowZero"
                      >
                        <div
                          class="euiFormControlLayout"
                        >
                          <div
                            class="euiFormControlLayout__childrenWrapper"
                          >
                            <input
                              class="euiFieldNumber"
                              data-test-subj="detectionInterval"
                              id="detectionInterval"
                              name="detectionInterval"
                              placeholder="Detector interval"
                              type="number"
                              value=""
                            />
                          </div>
                        </div>
                      </div>
                      <div
                        class="euiFlexItem"
                      >
                        <div
                          class="euiText euiText--medium"
                        >
                          <p
                            class="minutes"
                          >
                            minutes
                          </p>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="euiFlexItem"
              >
                <div
                  class="euiCheckbox"
                >
                  <input
                    checked=""
                    class="euiCheckbox__input"
                    id="runDetectorAutomaticallyCheckbox"
                    type="checkbox"
                  />
                  <div
                    class="euiCheckbox__square"
                  />
                  <label
                    class="euiCheckbox__label"
                    for="runDetectorAutomaticallyCheckbox"
                  >
                    Run detector automatically after creating
                  </label>
                </div>
              </div>
              <div
                class="euiFlexItem"
                style="margin-top: -8px;"
              >
                <p
                  class="sublabel"
                >
                   
                  We recommend running this detector during low traffic times.
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
